File: packages/oxlint-plugin-react-doctor/src/plugin/rules/react-native/rn-reanimated-4-no-legacy-spring-thresholds.ts
import { defineRule } from "../../utils/define-rule.js";
import type { EsTreeNodeOfType } from "../../utils/es-tree-node-of-type.js";
import { getStaticPropertyKeyName } from "../../utils/get-static-property-key-name.js";
import { isNodeOfType } from "../../utils/is-node-of-type.js";
import type { RuleContext } from "../../utils/rule-context.js";
import { stripParenExpression } from "../../utils/strip-paren-expression.js";
import { resolveReanimatedApiName } from "./utils/resolve-reanimated-api-name.js";

const WITH_SPRING_API_NAMES: ReadonlySet<string> = new Set(["withSpring"]);
const LEGACY_SPRING_THRESHOLD_NAMES: ReadonlySet<string> = new Set([
  "restDisplacementThreshold",
  "restSpeedThreshold",
]);

export const rnReanimated4NoLegacySpringThresholds = defineRule({
  id: "rn-reanimated-4-no-legacy-spring-thresholds",
  title: "Legacy Reanimated spring threshold",
  tags: ["migration-hint"],
  requires: ["reanimated:4"],
  severity: "warn",
  recommendation:
    "Replace Reanimated 3 rest thresholds with Reanimated 4's `energyThreshold` spring option.",
  create: (context: RuleContext) => ({
    CallExpression(node: EsTreeNodeOfType<"CallExpression">) {
      if (!resolveReanimatedApiName(node, context.scopes, WITH_SPRING_API_NAMES)) return;
      const configArgument = node.arguments[1];
      if (!configArgument || isNodeOfType(configArgument, "SpreadElement")) return;
      const unwrappedConfig = stripParenExpression(configArgument);
      if (!isNodeOfType(unwrappedConfig, "ObjectExpression")) return;

      for (const property of unwrappedConfig.properties) {
        if (!isNodeOfType(property, "Property")) continue;
        const propertyName = getStaticPropertyKeyName(property, {
          allowComputedString: true,
        });
        if (!propertyName || !LEGACY_SPRING_THRESHOLD_NAMES.has(propertyName)) continue;
        context.report({
          node: property,
          message: `Reanimated 4 removed \`${propertyName}\`; use the \`energyThreshold\` spring option instead.`,
        });
      }
    },
  }),
});

File: native/oxlint/rules/rn-reanimated-4-no-legacy-spring-thresholds.rs
use oxc_ast::{AstKind, ast::ObjectPropertyKind};
use oxc_diagnostics::OxcDiagnostic;
use oxc_macros::declare_oxc_lint;

use crate::{AstNode, context::LintContext, rule::Rule};

const REANIMATED_MODULE_SOURCES: [&str; 1] = ["react-native-reanimated"];
const LEGACY_SPRING_THRESHOLD_NAMES: [&str; 2] =
    ["restDisplacementThreshold", "restSpeedThreshold"];

#[derive(Debug, Default, Clone)]
#[allow(non_camel_case_types)]
pub struct RnReanimated_4NoLegacySpringThresholds;

pub type RnReanimated4NoLegacySpringThresholds = RnReanimated_4NoLegacySpringThresholds;

declare_oxc_lint!(
    /// Disallow legacy Reanimated spring thresholds.
    RnReanimated_4NoLegacySpringThresholds,
    react_doctor_native,
    correctness,
    version = "0.1.0",
    short_description = "Disallow legacy Reanimated spring thresholds.",
);

impl Rule for RnReanimated_4NoLegacySpringThresholds {
    fn run<'a>(&self, node: &AstNode<'a>, ctx: &LintContext<'a>) {
        let AstKind::CallExpression(call_expression) = node.kind() else {
            return;
        };
        if !module_api_path_matches(
            &call_expression.callee,
            &["withSpring"],
            &REANIMATED_MODULE_SOURCES,
            true,
            ctx,
        ) {
            return;
        }
        let Some(oxc_ast::ast::Expression::ObjectExpression(configuration)) = call_expression
            .arguments
            .get(1)
            .and_then(oxc_ast::ast::Argument::as_expression)
            .map(oxc_ast::ast::Expression::get_inner_expression)
        else {
            return;
        };
        for property in &configuration.properties {
            let ObjectPropertyKind::ObjectProperty(property) = property else {
                continue;
            };
            let Some(property_name) = LEGACY_SPRING_THRESHOLD_NAMES
                .iter()
                .find(|property_name| property_key_matches_name(&property.key, property_name))
            else {
                continue;
            };
            ctx.diagnostic(
                OxcDiagnostic::warn(format!(
                    "Reanimated 4 removed `{property_name}`; use the `energyThreshold` spring option instead."
                ))
                .with_label(property.span),
            );
        }
    }
}
